File: ruby/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary value to make testing easier
    # 1,024 would be way too high to test!
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
    raise 'Stack is full' if isFull?
    stack.push(item)
  end

  # remove item from top of stack and return it
  def pop
    @stack.pop
  end

  # return item at top of stack without removing it
  def peek
    @stack.last
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
    @stack.empty?
  end

  # return true if stack is full, otherwise false
  def isFull?
    @stack.size >= @limit
  end

  # return number of items in stack
  def size
    @stack.size
  end

  # return -1 if item not in stack, otherwise integer representing 
  # how far it is from the top
  def search(target)
    return -1 if @stack.index(target).nil?
    # 0 equals top or index -1
    # find the index of elem
    # subtract the stack size from the index number
    self.size - @stack.index(target) - 1
  end

  # print contents of stack: do not return the stack itself!
  def print
    puts @stack.join(' <- ')
  end
end

if __FILE__ == $PROGRAM_NAME
  # Don't forget to add your tests!
end

File: javascript/stack.js
class Stack {
  constructor() {
    this.stack = [];
    // this is an arbitrary value to make testing easier
    // 1,024 would be way too high to test!
    this.limit = 10;
  }

  // add item to top of stack if not full
  // if full throw error
  push(item) {
    if (this.isFull()) {
      throw new Error("Stack is full")
    }
    this.stack.push(item)
  }

  // remove item from top of stack and return it
  pop() {
    return this.stack.pop()
  }

  // return item at top of stack without removing it
  peek() {
    return this.stack[this.size - 1]
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
    return this.stack.length === 0
  }

  // return true if stack is full, otherwise false
  isFull() {
    if (this.stack.length >= this.limit) return true
    return false
  }

  // return number of items in stack
  size() {
    return this.stack.length
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top
  search(target) {
    if (this.stack.indexOf(target) === -1) return -1

    return this.size() - 1 - this.stack.indexOf(target)
  }

  // print contents of stack: do not return the stack itself!
  print() {
    console.log(this.stack.join(' <- '))
  }
}

if (require.main === module) {
  // add your own tests in here
}

module.exports = Stack;
